src/audio/asrc/asrc_farrow.c

```diff
@@ -551,52 +539,51 @@ enum asrc_error_code asrc_set_output_format(struct comp_dev *dev,
 /*
  * BUFFER FUNCTIONS
  */
-static enum asrc_error_code initialise_buffer(struct asrc_farrow *src_obj)
+static enum asrc_error_code initialise_buffer(struct comp_dev *dev,
+					      struct asrc_farrow *src_obj)
 {
-	uint8_t *buffer;
+	int32_t *start_32;
+	int16_t *start_16;
 	int ch;
-	int n;
-
-	/*
-	 * base_address points to the first address subsequently to the
-	 * memory where the pointers to each ring buffer are stored.
-	 */
-	if (src_obj->bit_depth == 32)
-		buffer = (uint8_t *)(src_obj->ring_buffers32 +
-			src_obj->num_channels);
-	else if (src_obj->bit_depth == 16)
-		buffer = (uint8_t *)(src_obj->ring_buffers16 +
-			src_obj->num_channels);
 
 	/*
-	 * set buffer_length to filter_length * 2 to compensate for
+	 * Set buffer_length to filter_length * 2 to compensate for
 	 * missing element wise wrap around while loading but allowing
 	 * aligned loads.
 	 */
 	src_obj->buffer_length = src_obj->filter_length * 2;
+	if (src_obj->buffer_length > ASRC_MAX_BUFFER_LENGTH) {
+		comp_err(dev, "initialise_buffer(), buffer_length %d exceeds max.",
+			 src_obj->buffer_length);
+		return ASRC_EC_INVALID_BUFFER_LENGTH;
+	}
+
 	src_obj->buffer_write_position = src_obj->filter_length;
 
-	/* set the base addresses for every channel and initialise the
-	 * buffers to zero
+	/*
+	 * Initialize the dynamically allocated 2D array and clear the
```

Comment:
not sure I'd call it a 2D array - it isn't one in C definition at least

Reply:
It's accessed in code as 2D array, see asrc_farrow_generic.c functions, like "src_obj->ring_buffers16[i][j];".

Reply:
Careful... Array of arrays isn't exactly 2D arrays when you store pointers themselves. Kinda weird legacy to get from the B language but that's that...

Reply:
Do you have recommendations how to change this before the major rewrite I'm thinking? I'd not like to start it now. To get your review verdict to a green check mark...

Reply:
I have already discussed the idea of having a linear memory region that would work both ways, plus an array of up to 8 pointers usable to get the 8 channels separate from that memory region. The allocation itself will be very interesting. @lyakh Do you think that's a good idea?

Reply:
Care to explain more, e.g. "both ways". What is it here more specifically that you don't like? As you see I'm not not a code wizard. At the moment I'm not interested in max. elegance but need code that can be trusted to work.

Reply:
@paulstelian97 sorry, could you explain a bit more? You don't seem to be proposing an interleaved buffer like @singalsu was proposing below, what kind of a linear region do you mean then?

Reply:
@lyakh Interleaved buffer would be best but I'm not sure this processing supports this. Currently the algorithm works with separate buffers for the differing channels. What I chimed in with was to allocate a single memory region for all the buffers for the channels (and I guess the coefficients as well)

If somehow the algorithm could be adapted to interleaved (and even support circular buffers), even better. But I'd be surprised if that happens.

Reply:
That's my planned next step after fixing the test fails.